src/Raven.Server/Documents/Indexes/Persistence/Lucene/OutputReduceIndexWriteOperation.cs

```diff
@@ -54,80 +55,101 @@ public override void IndexDocument(LazyStringValue key, object document, Indexin
-            public readonly List<OutputReduceDocument> ReduceDocuments = new List<OutputReduceDocument>();
+            private readonly List<OutputReduceDocument> _reduceDocuments = new List<OutputReduceDocument>();
+            private readonly JsonOperationContext _jsonContext;
 
             public OutputReduceToCollectionCommand(DocumentDatabase database, string outputReduceToCollection)
             {
                 _database = database;
                 _outputReduceToCollection = outputReduceToCollection;
+                _jsonContext = JsonOperationContext.ShortTermSingleUse();
             }
 
-            public class OutputReduceDocument
+            private class OutputReduceDocument
             {
                 public bool IsDelete;
-                public LazyStringValue ReduceKeyHash;
-                public object Document;
+                public string Key;
+                public BlittableJsonReaderObject Document;
             }
 
             public override void Execute(DocumentsOperationContext context)
             {
-                foreach (var reduceDocument in ReduceDocuments)
+                foreach (var reduceDocument in _reduceDocuments)
                 {
-                    var id = _outputReduceToCollection + "/" + reduceDocument.ReduceKeyHash;
-
+                    var key = reduceDocument.Key;
                     if (reduceDocument.IsDelete)
                     {
-                        _database.DocumentsStorage.Delete(context, id, null);
+                        _database.DocumentsStorage.Delete(context, key, null);
                         continue;
                     }
 
-                    var documentInfo = new DocumentInfo
-                    {
-                        Id = id,
-                        Collection = _outputReduceToCollection
-                    };
-                    using (var document = _entityToBlittable.ConvertEntityToBlittable(reduceDocument.Document, _documentConvention, context, documentInfo))
+                    using (var document = reduceDocument.Document)
                     {
-                        _database.DocumentsStorage.Put(context, id, null, document, flags: DocumentFlags.Artificial | DocumentFlags.FromIndex);
-                        context.DocumentDatabase.HugeDocuments.AddIfDocIsHuge(id, document.Size);
+                        _database.DocumentsStorage.Put(context, key, null, document, flags: DocumentFlags.Artificial | DocumentFlags.FromIndex);
+                        context.DocumentDatabase.HugeDocuments.AddIfDocIsHuge(key, document.Size);
                     }
                 }
             }
+
+            public void AddReduce(string reduceKeyHash, object reduceObject)
+            {
+                var key = _outputReduceToCollection + "/" + reduceKeyHash;
+
+                var djv = new DynamicJsonValue();
+                var propertyAccessor = PropertyAccessor.Create(reduceObject.GetType());
```

Comment:
Shouldn't this be cached? See the impl in `AnonymousLuceneDocumentConverter.Getfields`